huaweicloud/services/dds/resource_huaweicloud_dds_database_role.go

```diff
@@ -0,0 +1,326 @@
+		},
+
+		Importer: &schema.ResourceImporter{
+			StateContext: resourceDatabaseRoleImportState,
+		},
+
+		Schema: map[string]*schema.Schema{
+			"region": {
+				Type:     schema.TypeString,
+				Optional: true,
+				Computed: true,
+				ForceNew: true,
+			},
+			"instance_id": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"name": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+				ValidateFunc: validation.All(
+					validation.StringMatch(regexp.MustCompile(`^[\w-.]*$`),
+						"The name can only contain letters, digits, underscores (_), hyphens (-) and dots (.)."),
+					validation.StringLenBetween(1, 64),
+				),
+			},
+			"db_name": {
+				Type:     schema.TypeString,
+				Optional: true,
+				ForceNew: true,
+				ValidateFunc: validation.All(
+					validation.StringMatch(regexp.MustCompile(`^\w*$`),
+						"The name can only contain letters, digits and underscores (_)."),
+					validation.StringLenBetween(1, 64),
+				),
+				Default: "admin",
+			},
+			"roles": {
+				Type:     schema.TypeList,
+				Optional: true,
+				Computed: true,
+				ForceNew: true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"name": {
+							Type:     schema.TypeString,
+							Required: true,
+							ForceNew: true,
+						},
+						"db_name": {
+							Type:     schema.TypeString,
+							Optional: true,
+							ForceNew: true,
+							ValidateFunc: validation.All(
+								validation.StringMatch(regexp.MustCompile(`^\w*$`),
+									"The name can only contain letters, digits and underscores (_)."),
+								validation.StringLenBetween(1, 64),
+							),
+							Default: "admin",
+						},
+					},
+				},
+			},
+			"privileges": {
+				Type:     schema.TypeList,
+				Computed: true,
+				Elem:     privilegeSchemaResource(),
+			},
+			"inherited_privileges": {
+				Type:     schema.TypeList,
+				Computed: true,
+				Elem:     privilegeSchemaResource(),
+			},
+		},
+	}
+}
+
+func buildDatabaseRoles(roleList []interface{}) []roles.Role {
+	result := make([]roles.Role, len(roleList))
+	for i, val := range roleList {
+		role := val.(map[string]interface{})
+		result[i] = roles.Role{
+			Name:   role["name"].(string),
+			DbName: role["db_name"].(string),
+		}
+	}
+	return result
+}
+
+func buildCreateOpts(d *schema.ResourceData) roles.CreateOpts {
+	return roles.CreateOpts{
+		Name:   d.Get("name").(string),
+		DbName: d.Get("db_name").(string),
+		Roles:  buildDatabaseRoles(d.Get("roles").([]interface{})),
+	}
+}
+
+func resourceDatabaseRoleCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	conf := meta.(*config.Config)
+	client, err := conf.DdsV3Client(conf.GetRegion(d))
+	if err != nil {
+		return diag.Errorf("error creating DDS v3 client: %s", err)
+	}
+
+	instanceId := d.Get("instance_id").(string)
+	config.MutexKV.Lock(instanceId)
+	defer config.MutexKV.Unlock(instanceId)
```

Comment:
this part looks not necessary

Reply:
This lock is in order to avoid the concurrency problem of database operations.